### PR TITLE
Adding unused local assignment check to compiler strict mode

### PR DESCRIPTION
**Pull request by johanfylling**
Fixes: #2514
Signed-off-by: Johan Fylling <johan.dev@fylling.se>

### CODE REVIEW

File: ast/compile.go [hunk 1]
```diff
@@ -4165,66 +4232,84 @@ func rewriteDeclaredVarsInTerm(g *localVarGenerator, stack *localDeclaredVars, t
...
 	})
 	return errs
 }
 
-func rewriteDeclaredVarsInArrayComprehension(g *localVarGenerator, stack *localDeclaredVars, v *ArrayComprehension, errs Errors) Errors {
+func rewriteDeclaredVarsInArrayComprehension(g *localVarGenerator, stack *localDeclaredVars, v *ArrayComprehension, errs Errors, strict bool) Errors {
+	used := NewVarSet()
+	if v.Term != nil {
+		used.Update(v.Term.Vars())
+	}
```
Comment (johanfylling) [inline ast/compile.go:4278]: This will change the error generated by `p := [i | some i]` from `declared var i unused` to `var i is unsafe`.
set and object comprehension is similarly changed.

File: ast/compile.go [hunk 2]
```diff
@@ -4017,7 +4083,7 @@ func checkUnusedDeclaredVars(loc *Location, stack *localDeclaredVars, used VarSe
 	return errs
 }
 
-func rewriteSomeDeclStatement(g *localVarGenerator, stack *localDeclaredVars, expr *Expr, errs Errors) (*Expr, Errors) {
+func rewriteSomeDeclStatement(g *localVarGenerator, stack *localDeclaredVars, expr *Expr, errs Errors, strict bool) (*Expr, Errors) {
```
Comment (johanfylling) [inline ast/compile.go:4086]: I ended up sprinkling these `strict` args in a lot of places ... If they're an annoyance, I could find another solution. Would it be better to attach all these functions to the compiler?

File: ast/compile.go [hunk 3]
```diff
@@ -1806,7 +1808,9 @@ func (vis *ruleArgLocalRewriter) Visit(x interface{}) Visitor {
 	switch v := t.Value.(type) {
 	case Var:
 		gv, ok := vis.stack.Declared(v)
-		if !ok {
+		if ok {
+			vis.stack.Seen(v)
```
Comment (johanfylling) [inline ast/compile.go:1812]: This will only be called when rule args are traversed, and I believe there there is no way of assigning to such vars without getting a `redeclared` error (correct me if I'm wrong). So in essence, incrementing the counter for a seen var here is irrelevant for this feature; but I added this here fore the sake of consistency.

File: ast/compile.go [hunk 4]
```diff
@@ -3937,6 +3945,30 @@ func (s localDeclaredVars) GlobalOccurrence(x Var) (varOccurrence, bool) {
 	return newVar, false
 }
 
+// Seen marks x as seen by incrementing its counter
+func (s localDeclaredVars) Seen(x Var) {
```
Comment (jcchavezs) [inline ast/compile.go:3949]: I guess this can't be called concurrently but just wondering.

File: .gitignore [hunk 5]
```diff
@@ -20,6 +20,7 @@ policy.wasm
 .go
 release-notes.md
 bundle.tar.gz
+/docs/website/.hugo_build.lock
```
Comment (srenatus) [inline .gitignore:23]: [nit] Looks unrelated, can we drop this? 😉

File: ast/compile.go (hunk 2 above)
Comment (srenatus) [reply ast/compile.go:4086]: I think that's OK, and consistent with the prevalent style of the compiler code... 🤔

File: ast/compile.go (hunk 3 above)
Comment (srenatus) [reply ast/compile.go:1812]: Just thinking aloud about how this branch could be triggered, perhaps
```rego
f(x, x) { true }
```
So when the second `x` is visited, it will already have been deal with...?

File: ast/compile.go [hunk 6]
```diff
@@ -3978,9 +4010,43 @@ func rewriteDeclaredVarsInBody(g *localVarGenerator, stack *localDeclaredVars, u
...
+
+	if !strict || len(errs) > 0 {
+		return errs
+	}
+
+	dvs := stack.Peek()
+	unused := NewVarSet()
+
+	for v, occ := range dvs.occurrence {
+		if !v.IsWildcard() && occ == assignedVar && stack.Count(v) <= 1 {
```
Comment (srenatus) [inline ast/compile.go:4027]: Do we need to treat "count == 1" as unused..? 🤔 that will perhaps some day in the future surprise me a little.

File: ast/compile.go (hunk 1 above)
Comment (srenatus) [reply ast/compile.go:4278]: Can these keys, and the object pairs, ever be nil? I might just be missing something 🤔

**Review by srenatus — COMMENTED** [review]
A few comments, but this looks pretty good! 😃

File: .gitignore (hunk 5 above)
Comment (johanfylling) [reply .gitignore:23]: Sure. This one had been bugging me, as it always shows up when I build the docs locally to check that my changes look ok 😅 .

File: ast/compile.go (hunk 4 above)
Comment (johanfylling) [reply ast/compile.go:3949]: In this case, I don't think that would improve performance enough to warrant the added complexity.

File: ast/compile.go (hunk 6 above)
Comment (johanfylling) [reply ast/compile.go:4027]: I can add a comment here. Since the var was assigned in this scope, it also needs to have been seen more than once (time of assignment) in this, and nested, scopes.

Comment (srenatus) [reply ast/compile.go:4027]: Ah, assignment is considered use, of course. That makes sense, I was too caught up thinking about _declarations_ instead of _assignments_. 👍

File: ast/compile.go (hunk 1 above)
Comment (johanfylling) [reply ast/compile.go:4278]: I don't think so, just playing it safe.

Comment (johanfylling) [reply ast/compile.go:4278]: I don't think so, just playing it safe.

File: ast/compile.go (hunk 3 above)
Comment (johanfylling) [reply ast/compile.go:1812]: Yes, that would do it. But I don't think there is ever a case that would trigger here that applies to the problem of unused local var assignments (might be wrong, though; at which case I should add a test case for it). But it's good to be consistent, I think. Who knows if this is useful fore something else in the future.

File: ast/compile_test.go [hunk 7]
```diff
@@ -3341,6 +3341,376 @@ func TestCompilerMockVirtualDocumentPartially(t *testing.T) {
 	assertCompilerErrorStrings(t, c, []string{"rego_compile_error: with keyword cannot partially replace virtual document(s)"})
 }
 
+func TestCompilerCheckUnusedAssignedVar(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (srenatus) [reply ast/compile_test.go:3344]: it's not _glaring_, but perhaps we could have a nested closure case?
```
p { 
  x := 1
  { y | y := [ z | z:=[1,2,3][_]; z > 1 ][_] }
}
``` 
though I'm not sure right now how this would look after the previous compiler stages have processed it 🤔

File: ast/compile.go (hunk 4 above)
Comment (srenatus) [reply ast/compile.go:3949]: The compiler runs everything sequentially, so it's not concurrency-safe in many places, I'm afraid.

File: ast/compile.go (hunk 1 above)
Comment (srenatus) [reply ast/compile.go:4280]: I think this is muddying the waters a bit. But it's not a blocker, of course.

Comment (johanfylling) [reply ast/compile.go]: Did some extra checking, and indeed it looks safe to not do any nil-checking.

**Review by srenatus — APPROVED** [review]
